src/server/scripts/Spells/spell_generic.cpp

```diff
@@ -1614,6 +1614,74 @@ class spell_gen_elune_candle : public SpellScript
     }
 };
 
+// 50051 - Ethereal Pet Aura
+enum EtherealPet
+{
+    SAY_STEAL_ESSENCE           = 1,
+    SAY_CREATE_TOKEN            = 2,
+
+    SPELL_STEAL_ESSENCE_VISUAL  = 50101,
+    SPELL_CREATE_TOKEN          = 50063
+};
+
+class spell_ethereal_pet_aura : public AuraScript
+{
+    PrepareAuraScript(spell_ethereal_pet_aura);
+
+    bool CheckProc(ProcEventInfo& eventInfo)
+    {
+        uint32 levelDiff = std::abs(GetTarget()->getLevel() - eventInfo.GetProcTarget()->getLevel());
+        if (levelDiff >= 9)
+            return false;
+
+        return true;
+    }
+
+    void OnProc(AuraEffect const* /*aurEff*/, ProcEventInfo& eventInfo)
+    {
+        PreventDefaultAction();
+
+        Player* owner = GetTarget()->ToPlayer();
+        Unit* target = eventInfo.GetActionTarget();
+        if (!owner || !target)
+            return;
+
+        Creature* pet = owner->GetMap()->GetCreature(owner->m_SummonSlot[SUMMON_TYPE_MINIPET]);
+        if (!pet)
+            return;
+
+        pet->CastSpell(target, SPELL_STEAL_ESSENCE_VISUAL);
+        pet->AI()->Talk(SAY_STEAL_ESSENCE);
+    }
+
+    void Register() override
+    {
+        DoCheckProc += AuraCheckProcFn(spell_ethereal_pet_aura::CheckProc);
+        OnEffectProc += AuraEffectProcFn(spell_ethereal_pet_aura::OnProc, EFFECT_0, SPELL_AURA_PROC_TRIGGER_SPELL);
+    }
+};
+
+// 50101 - Steal Essence Visual
+class spell_steal_essence_visual : public AuraScript
+{
+    PrepareAuraScript(spell_steal_essence_visual);
+
+    void HandleRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        if (!GetCaster() || !GetCaster()->GetOwner())
+            return;
+
+        Creature* pet = GetCaster()->ToCreature();
```

Comment:
please @DanVS take the required time to review all the required checks in this function and to understand why they are needed.
When removing an Aura from a Target, the Caster could have already logged out so that's why we need to NULL-check GetCaster().
When we cast Caster to a Creature with ToCreature(), there is no guarantee it actually is a Creature and in case of a Player it will return NULL, requiring a NULL check here too on the "pet" variable (which you added in a previous commit but then removed).
Thanks for the PR :)

Reply:
Thanks, jackpoz! Do I understand correctly that previous version was fine? If `GetCaster()` returns NULL, but we have casted it to Creature immediately than NULL will be returned anyway? So we can do check shortly after variables declaration?

Reply:
OK, I understood mistake of mine. Thanks a lot :)